File: sources/net.sf.j2s.core/test/src/com/falstad/DotProduct.java
// DotProduct.java (C) 2001 by Paul Falstad, www.falstad.com

// Added imports
//
// paint --> paintComponent

package com.falstad;
import java.awt.Color;
import java.awt.Component;
import java.awt.Container;
import java.awt.Dimension;
import java.awt.FontMetrics;
import java.awt.Font;
import java.awt.Graphics;
import java.awt.Image;
import java.awt.LayoutManager;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.AdjustmentEvent;
import java.awt.event.AdjustmentListener;
import java.awt.event.ComponentEvent;
import java.awt.event.ComponentListener;
import java.awt.event.MouseEvent;
import java.awt.event.MouseListener;
import java.awt.event.MouseMotionListener;
import java.text.NumberFormat;
import java.util.Random;

import a2s.Applet;

import a2s.Canvas;
import a2s.Button;

//web_Ready
//web_AppletName= DotProduct
//web_Description= Demonstrates the dot product or scalar product of two vectors
//web_JavaVersion= http://www.falstad.com/dotproduct/
//web_AppletImage= dotproduct.png
//web_Category= Mathematics
//web_Date= $Date: 2016-12-31 17:53:43 -0600 (Sat, 31 Dec 2016) $
//web_Features: graphics, AWT-to-Swing

class DotProductCanvas extends Canvas {
 DotProduct pg;
 DotProductCanvas(DotProduct p) {
	pg = p;
	setBackground(Color.BLACK);
 }
 public Dimension getPreferredSize() {
	return new Dimension(300,400);
 }
 public void update(Graphics g) {
	pg.updateDotProduct(g);
 }
 public void paintComponent(Graphics g) {
	 super.paintComponent(g);
	pg.updateDotProduct(g);
 }
};

class DotProductLayout implements LayoutManager {
 public DotProductLayout() {}
 public void addLayoutComponent(String name, Component c) {}
 public void removeLayoutComponent(Component c) {}
 public Dimension preferredLayoutSize(Container target) {
	return new Dimension(500, 500);
 }
 public Dimension minimumLayoutSize(Container target) {
	return new Dimension(100,100);
 }
 public void layoutContainer(Container target) {
	int cw = target.size().width;
	int ct = target.getComponentCount();
	target.getComponent(ct-1).move(0, 0);
	target.getComponent(ct-1).resize(cw, target.size().height);
	int i;
	int h = 0;
	for (i = 0; i < ct-1; i++) {
	    Component m = target.getComponent(i);
	    if (m.isVisible()) {
		Dimension d = m.getPreferredSize();
		m.move(cw-d.width, h);
		m.resize(d.width, d.height);
		h += d.height;
	    }
	}
 }
};


public class DotProduct extends Applet
implements ComponentListener, ActionListener, AdjustmentListener,
          MouseMotionListener, MouseListener {
 
 Thread engine = null;

 Dimension winSize;
 Image dbimage;
 
 Random random;
 
 public String getAppletInfo() {
	return "DotProduct by Paul Falstad";
 }

 Button swapButton;
 double vecs[][];
 int selection = -1;

 int getrand(int x) {
	int q = random.nextInt();
	if (q < 0) q = -q;
	return q % x;
 }
 DotProductCanvas cv;

 public void init() {
	setLayout(new DotProductLayout());
	cv = new DotProductCanvas(this);
	cv.addComponentListener(this);
	cv.addMouseMotionListener(this);
	cv.addMouseListener(this);
	add(swapButton = new Button("Swap"));
	swapButton.addActionListener(this);
	add(cv);
	setBackground(Color.black);
	setForeground(Color.lightGray);
	random = new Random();
	vecs = new double[2][2];
	vecs[0][0] = 0; vecs[0][1] = 1;
	vecs[1][0] = 1; vecs[1][1] = 1;
	reinit();
	repaint();
 }

 void reinit() {
     Dimension d = winSize = cv.getSize();
	if (winSize.width == 0)
	    return;
	dbimage = createImage(d.width, d.height);
 }
 
 /*
 public void paint(Graphics g) {
	cv.repaint();
 }
*/
 
 void findVecCoords(double x, double y, int result[]) {
	int cy = winSize.height/4;
	int cx = cy;
	result[0] = (int) (cx*(x+2));
	result[1] = (int) (cy*(2-y));
 }

 void findVecCoords(int num, int result[]) {
	findVecCoords(vecs[num][0], vecs[num][1], result);
 }

 void drawArrow(Graphics g, int x1, int y1, int x2, int y2, double len) {
	g.drawLine(x1, y1, x2, y2);
	if (len > .05) {
	    double l = java.lang.Math.sqrt((x2-x1)*(x2-x1)+(y2-y1)*(y2-y1));
	    double hatx = (x2-x1)/l;
	    double haty = (y2-y1)/l;
	    int as = 10;
	    g.drawLine(x2, y2,
		       (int) (haty*as-hatx*as+x2),
		       (int) (-hatx*as-haty*as+y2));
	    g.drawLine(x2, y2,
		       (int) (-haty*as-hatx*as+x2),
		       (int) (hatx*as-haty*as+y2));
	}
 }

 void drawBar(Graphics g, int offset, double val) {
	int x = (int) (winSize.width * val/6);
	int cx = winSize.width/2;
	int h = 5;
	int y = winSize.height + h*offset;
	int y2 = y+h-1;
	if (val < 0)
	    g.fillRect(cx+x, y, -x, h);
	else
	    g.fillRect(cx, y, x, h);
 }

 public void updateDotProduct(Graphics realg) {
	double alen = java.lang.Math.sqrt(
	    vecs[0][0] * vecs[0][0] + vecs[0][1] * vecs[0][1]);
	double blen = java.lang.Math.sqrt(
	    vecs[1][0] * vecs[1][0] + vecs[1][1] * vecs[1][1]);
	double piadj = 180/3.14159265;
	double dot = vecs[0][0] * vecs[1][0] + vecs[0][1] * vecs[1][1];
	double acosth = (blen > 0) ? dot/blen : 0;
	double costh = (alen > 0) ? acosth/alen : 0;
	double theta = java.lang.Math.acos(costh)*piadj;

	Graphics g = dbimage.getGraphics();
	if (winSize == null || winSize.width == 0)
	    return;
	g.setColor(cv.getBackground());
	g.fillRect(0, 0, winSize.width, winSize.height);
	g.setColor(Color.gray);
	g.setFont(new Font("Helvetica", 0, 15));
	int i, j;
	for (i = -2; i <= 2; i++) {
	    int x = winSize.height*(i+2)/4;
	    g.drawLine(x, 0, x, winSize.height);
	    g.drawLine(0, x, winSize.height, x);
	}
	int cy = winSize.height/2;
	int cx = cy;
	int vc[] = new int[2];
	if (blen > 0) {
	    int vc2[] = new int[2];
	    findVecCoords(vecs[1][0] * acosth/blen,
			  vecs[1][1] * acosth/blen, vc);
	    findVecCoords(0, vc2);
	    g.setColor(Color.gray);
	    g.drawLine(vc[0], vc[1], vc2[0], vc2[1]);
	}
	if (alen > .1 && blen > .1) {
	    int c1x = cx/10;
	    int c1y = cy/10;
	    int a1 = (int)(piadj*java.lang.Math.atan2(vecs[0][1], vecs[0][0]));
	    int a2 = (int)(piadj*java.lang.Math.atan2(vecs[1][1], vecs[1][0]));
	    if (a1 > a2 && a1 < a2+180) {
		int a3 = a1;
		a1 = a2; 
		a2 = a3;
	    }
	    if (a2 < a1)
		a2 += 360;
	    g.setColor(Color.orange);
	    g.drawArc(cx-c1x, cy-c1y, c1x*2, c1y*2, a1, a2-a1);
	}
	findVecCoords(0, vc);
	g.setColor(Color.red);
	drawArrow(g, cx, cy, vc[0], vc[1], alen);
	findVecCoords(1, vc);
	g.setColor(Color.cyan);
	drawArrow(g, cx, cy, vc[0], vc[1], blen);
	int yl = g.getFontMetrics().getHeight();
	int y = yl;
	NumberFormat nf = NumberFormat.getInstance();
	nf.setMaximumFractionDigits(3);
	g.setColor(Color.red);
	displayString(g,
		     "A = (" + nf.format(vecs[0][0]) + ", " +
		     nf.format(vecs[0][1]) + ")", y += yl);
	displayString(g, "|A| = " + nf.format(alen), y += yl);
	drawBar(g, -4, alen);
	g.setColor(Color.cyan);
	displayString(g,
		     "B = (" + nf.format(vecs[1][0]) + ", " +
		     nf.format(vecs[1][1]) + ")", y += yl);
	displayString(g, "|B| = " + nf.format(blen), y += yl);
	drawBar(g, -3, blen);
	g.setColor(Color.yellow);
	displayString(g, "|A| cos theta = " + nf.format(acosth), y += yl);
	drawBar(g, -2, acosth);
	if (blen > 0) {
	    findVecCoords(vecs[1][0] * acosth/blen,
			  vecs[1][1] * acosth/blen, vc);
	    g.setColor(Color.yellow);
	    g.drawLine(cx, cy, vc[0], vc[1]);
	}
	g.setColor(Color.white);
	displayString(g, "cos theta = " + nf.format(costh), y += yl);
	g.setColor(Color.orange);
	displayString(g, "theta = " + nf.format(theta) + "\u00b0", y += yl);
	g.setColor(Color.green);
	displayString(g, "A dot B = " + nf.format(dot), y += yl);
	drawBar(g, -1, dot);
	realg.drawImage(dbimage, 0, 0, this);
 }

 void displayString(Graphics g, String s, int y) {
	int lx = winSize.height;
	int lw = winSize.width - lx;
	FontMetrics fm = g.getFontMetrics();
     g.drawString(s, lx+(lw-fm.stringWidth(s))/2, y);
 }

 void edit(MouseEvent e) {
	if (selection == -1)
	    return;
	int x = e.getX();
	int y = e.getY();
	double cy = winSize.height/4;
	double cx = cy;
	double xf = x/cx-2;
	double yf = 2-y/cy;
	if (xf < -2) xf = -2;
	if (yf < -2) yf = -2;
	if (xf >  2) xf =  2;
	if (yf >  2) yf =  2;
	vecs[selection][0] = xf;
	vecs[selection][1] = yf;
	repaint();
 }

 public void componentHidden(ComponentEvent e){}
 public void componentMoved(ComponentEvent e){}
 public void componentShown(ComponentEvent e) {
	repaint();
 }

 public void componentResized(ComponentEvent e) {
	reinit();
	repaint();
 }
 public void actionPerformed(ActionEvent e) {
	if (e.getSource() == swapButton) {
	    int x;
	    for (x = 0; x < 2; x++) {
		double y = vecs[0][x];
		vecs[0][x] = vecs[1][x];
		vecs[1][x] = y;
	    }
	    repaint();
	}
 }
 public void adjustmentValueChanged(AdjustmentEvent e) {
 }
 public void mouseDragged(MouseEvent e) {
	edit(e);
 }
 public void mouseMoved(MouseEvent e) {
	if ((e.getModifiers() & MouseEvent.BUTTON1_MASK) != 0)
	    return;
	edit(e);
 }
 public void mouseClicked(MouseEvent e) {
 }
 public void mouseEntered(MouseEvent e) {
 }
 public void mouseExited(MouseEvent e) {
 }
 public void mousePressed(MouseEvent e) {
	if ((e.getModifiers() & MouseEvent.BUTTON1_MASK) == 0)
	    return;
	int x = e.getX();
	int y = e.getY();
	int vc[] = new int[2];
	int i;
	selection = -1;
	float best = 30*30;
	for (i = 0; i != 2; i++) {
	    findVecCoords(i, vc);
	    float dx = x-vc[0];
	    float dy = y-vc[1];
	    float dist = dx*dx+dy*dy;
	    if (dist < best) {
	    	best = dist;
	    	selection = i;
	    }
	}
	if (selection != -1)
	    edit(e);
 }
 public void mouseReleased(MouseEvent e) {
	if ((e.getModifiers() & MouseEvent.BUTTON1_MASK) == 0)
	    return;
	selection = -1;
 }
}

File: sources/net.sf.j2s.core/test/src/com/falstad/DotProduct.js
(function(){var P$=Clazz.newPackage("com.falstad"),I$=[];
var C$=Clazz.newClass(P$, "DotProduct", null, 'a2s.Applet', ['java.awt.event.ComponentListener', 'java.awt.event.ActionListener', 'java.awt.event.AdjustmentListener', 'java.awt.event.MouseMotionListener', 'java.awt.event.MouseListener']);

C$.$clinit$ = function() {Clazz.load(C$, 1);
}

Clazz.newMeth(C$, '$init0$', function () {
var c;if((c = C$.superclazz) && (c = c.$init0$))c.apply(this);
this.engine = null;
this.winSize = null;
this.dbimage = null;
this.random = null;
this.swapButton = null;
this.vecs = null;
this.selection = 0;
this.cv = null;
}, 1);

Clazz.newMeth(C$, '$init$', function () {
this.engine = null;
this.selection = -1;
}, 1);

Clazz.newMeth(C$, 'getAppletInfo', function () {
return "DotProduct by Paul Falstad";
});

Clazz.newMeth(C$, 'getrand$I', function (x) {
var q = this.random.nextInt();
if (q < 0) q = -q;
return q % x;
});

Clazz.newMeth(C$, 'init', function () {
this.setLayout$java_awt_LayoutManager(Clazz.new_((I$[2]||(I$[2]=Clazz.load('com.falstad.DotProductLayout')))));
this.cv = Clazz.new_((I$[3]||(I$[3]=Clazz.load('com.falstad.DotProductCanvas'))).c$$com_falstad_DotProduct,[this]);
this.cv.addComponentListener$java_awt_event_ComponentListener(this);
this.cv.addMouseMotionListener$java_awt_event_MouseMotionListener(this);
this.cv.addMouseListener$java_awt_event_MouseListener(this);
this.add$java_awt_Component(this.swapButton = Clazz.new_((I$[4]||(I$[4]=Clazz.load('a2s.Button'))).c$$S,["Swap"]));
this.swapButton.addActionListener$java_awt_event_ActionListener(this);
this.add$java_awt_Component(this.cv);
this.setBackground$java_awt_Color((I$[0]||(I$[0]=Clazz.load('java.awt.Color'))).black);
this.setForeground$java_awt_Color((I$[0]||(I$[0]=Clazz.load('java.awt.Color'))).lightGray);
this.random = Clazz.new_((I$[5]||(I$[5]=Clazz.load('java.util.Random'))));
this.vecs = Clazz.array(Double.TYPE, [2, 2]);
this.vecs[0][0] = 0;
this.vecs[0][1] = 1;
this.vecs[1][0] = 1;
this.vecs[1][1] = 1;
this.reinit();
this.repaint();
});

Clazz.newMeth(C$, 'reinit', function () {
var d = this.winSize = this.cv.getSize();
if (this.winSize.width == 0) return;
this.dbimage = this.createImage$I$I(d.width, d.height);
});

Clazz.newMeth(C$, 'findVecCoords$D$D$IA', function (x, y, result) {
var cy = (this.winSize.height/4|0);
var cx = cy;
result[0] = ((cx * (x + 2))|0);
result[1] = ((cy * (2 - y))|0);
});

Clazz.newMeth(C$, 'findVecCoords$I$IA', function (num, result) {
this.findVecCoords$D$D$IA(this.vecs[num][0], this.vecs[num][1], result);
});

Clazz.newMeth(C$, 'drawArrow$java_awt_Graphics$I$I$I$I$D', function (g, x1, y1, x2, y2, len) {
g.drawLine$I$I$I$I(x1, y1, x2, y2);
if (len > 0.05 ) {
var l = java.lang.Math.sqrt((x2 - x1) * (x2 - x1) + (y2 - y1) * (y2 - y1));
var hatx = (x2 - x1) / l;
var haty = (y2 - y1) / l;
var as = 10;
g.drawLine$I$I$I$I(x2, y2, ((haty * as - hatx * as + x2)|0), ((-hatx * as - haty * as + y2)|0));
g.drawLine$I$I$I$I(x2, y2, ((-haty * as - hatx * as + x2)|0), ((hatx * as - haty * as + y2)|0));
}});

Clazz.newMeth(C$, 'drawBar$java_awt_Graphics$I$D', function (g, offset, val) {
var x = ((this.winSize.width * val / 6)|0);
var cx = (this.winSize.width/2|0);
var h = 5;
var y = this.winSize.height + h * offset;
var y2 = y + h - 1;
if (val < 0 ) g.fillRect$I$I$I$I(cx + x, y, -x, h);
 else g.fillRect$I$I$I$I(cx, y, x, h);
});

Clazz.newMeth(C$, 'updateDotProduct$java_awt_Graphics', function (realg) {
var alen = java.lang.Math.sqrt(this.vecs[0][0] * this.vecs[0][0] + this.vecs[0][1] * this.vecs[0][1]);
var blen = java.lang.Math.sqrt(this.vecs[1][0] * this.vecs[1][0] + this.vecs[1][1] * this.vecs[1][1]);
var piadj = 57.29577957855229;
var dot = this.vecs[0][0] * this.vecs[1][0] + this.vecs[0][1] * this.vecs[1][1];
var acosth = (blen > 0 ) ? dot / blen : 0;
var costh = (alen > 0 ) ? acosth / alen : 0;
var theta = java.lang.Math.acos(costh) * piadj;
var g = this.dbimage.getGraphics();
if (this.winSize == null  || this.winSize.width == 0 ) return;
g.setColor$java_awt_Color(this.cv.getBackground());
g.fillRect$I$I$I$I(0, 0, this.winSize.width, this.winSize.height);
g.setColor$java_awt_Color((I$[0]||(I$[0]=Clazz.load('java.awt.Color'))).gray);
g.setFont$java_awt_Font(Clazz.new_((I$[6]||(I$[6]=Clazz.load('java.awt.Font'))).c$$S$I$I,["Helvetica", 0, 15]));
var i;
var j;
for (i = -2; i <= 2; i++) {
var x = (this.winSize.height * (i + 2)/4|0);
g.drawLine$I$I$I$I(x, 0, x, this.winSize.height);
g.drawLine$I$I$I$I(0, x, this.winSize.height, x);
}
var cy = (this.winSize.height/2|0);
var cx = cy;
var vc = Clazz.array(Integer.TYPE, [2]);
if (blen > 0 ) {
var vc2 = Clazz.array(Integer.TYPE, [2]);
this.findVecCoords$D$D$IA(this.vecs[1][0] * acosth / blen, this.vecs[1][1] * acosth / blen, vc);
this.findVecCoords$I$IA(0, vc2);
g.setColor$java_awt_Color((I$[0]||(I$[0]=Clazz.load('java.awt.Color'))).gray);
g.drawLine$I$I$I$I(vc[0], vc[1], vc2[0], vc2[1]);
}if (alen > 0.1  && blen > 0.1  ) {
var c1x = (cx/10|0);
var c1y = (cy/10|0);
var a1 = ((piadj * java.lang.Math.atan2(this.vecs[0][1], this.vecs[0][0]))|0);
var a2 = ((piadj * java.lang.Math.atan2(this.vecs[1][1], this.vecs[1][0]))|0);
if (a1 > a2 && a1 < a2 + 180 ) {
var a3 = a1;
a1 = a2;
a2 = a3;
}if (a2 < a1) a2 = a2+(360);
g.setColor$java_awt_Color((I$[0]||(I$[0]=Clazz.load('java.awt.Color'))).orange);
g.drawArc$I$I$I$I$I$I(cx - c1x, cy - c1y, c1x * 2, c1y * 2, a1, a2 - a1);
}this.findVecCoords$I$IA(0, vc);
g.setColor$java_awt_Color((I$[0]||(I$[0]=Clazz.load('java.awt.Color'))).red);
this.drawArrow$java_awt_Graphics$I$I$I$I$D(g, cx, cy, vc[0], vc[1], alen);
this.findVecCoords$I$IA(1, vc);
g.setColor$java_awt_Color((I$[0]||(I$[0]=Clazz.load('java.awt.Color'))).cyan);
this.drawArrow$java_awt_Graphics$I$I$I$I$D(g, cx, cy, vc[0], vc[1], blen);
var yl = g.getFontMetrics().getHeight();
var y = yl;
var nf = (I$[7]||(I$[7]=Clazz.load('java.text.NumberFormat'))).getInstance();
nf.setMaximumFractionDigits$I(3);
g.setColor$java_awt_Color((I$[0]||(I$[0]=Clazz.load('java.awt.Color'))).red);
this.displayString$java_awt_Graphics$S$I(g, "A = (" + nf.format$D(this.vecs[0][0]) + ", " + nf.format$D(this.vecs[0][1]) + ")" , y = y+(yl));
this.displayString$java_awt_Graphics$S$I(g, "|A| = " + nf.format$D(alen), y = y+(yl));
this.drawBar$java_awt_Graphics$I$D(g, -4, alen);
g.setColor$java_awt_Color((I$[0]||(I$[0]=Clazz.load('java.awt.Color'))).cyan);
this.displayString$java_awt_Graphics$S$I(g, "B = (" + nf.format$D(this.vecs[1][0]) + ", " + nf.format$D(this.vecs[1][1]) + ")" , y = y+(yl));
this.displayString$java_awt_Graphics$S$I(g, "|B| = " + nf.format$D(blen), y = y+(yl));
this.drawBar$java_awt_Graphics$I$D(g, -3, blen);
g.setColor$java_awt_Color((I$[0]||(I$[0]=Clazz.load('java.awt.Color'))).yellow);
this.displayString$java_awt_Graphics$S$I(g, "|A| cos theta = " + nf.format$D(acosth), y = y+(yl));
this.drawBar$java_awt_Graphics$I$D(g, -2, acosth);
if (blen > 0 ) {
this.findVecCoords$D$D$IA(this.vecs[1][0] * acosth / blen, this.vecs[1][1] * acosth / blen, vc);
g.setColor$java_awt_Color((I$[0]||(I$[0]=Clazz.load('java.awt.Color'))).yellow);
g.drawLine$I$I$I$I(cx, cy, vc[0], vc[1]);
}g.setColor$java_awt_Color((I$[0]||(I$[0]=Clazz.load('java.awt.Color'))).white);
this.displayString$java_awt_Graphics$S$I(g, "cos theta = " + nf.format$D(costh), y = y+(yl));
g.setColor$java_awt_Color((I$[0]||(I$[0]=Clazz.load('java.awt.Color'))).orange);
this.displayString$java_awt_Graphics$S$I(g, "theta = " + nf.format$D(theta) + "\u00b0" , y = y+(yl));
g.setColor$java_awt_Color((I$[0]||(I$[0]=Clazz.load('java.awt.Color'))).green);
this.displayString$java_awt_Graphics$S$I(g, "A dot B = " + nf.format$D(dot), y = y+(yl));
this.drawBar$java_awt_Graphics$I$D(g, -1, dot);
realg.drawImage$java_awt_Image$I$I$java_awt_image_ImageObserver(this.dbimage, 0, 0, this);
});

Clazz.newMeth(C$, 'displayString$java_awt_Graphics$S$I', function (g, s, y) {
var lx = this.winSize.height;
var lw = this.winSize.width - lx;
var fm = g.getFontMetrics();
g.drawString$S$I$I(s, lx + ((lw - fm.stringWidth$S(s))/2|0), y);
});

Clazz.newMeth(C$, 'edit$java_awt_event_MouseEvent', function (e) {
if (this.selection == -1) return;
var x = e.getX();
var y = e.getY();
var cy = (this.winSize.height/4|0);
var cx = cy;
var xf = x / cx - 2;
var yf = 2 - y / cy;
if (xf < -2 ) xf = -2;
if (yf < -2 ) yf = -2;
if (xf > 2 ) xf = 2;
if (yf > 2 ) yf = 2;
this.vecs[this.selection][0] = xf;
this.vecs[this.selection][1] = yf;
this.repaint();
});

Clazz.newMeth(C$, 'componentHidden$java_awt_event_ComponentEvent', function (e) {
});

Clazz.newMeth(C$, 'componentMoved$java_awt_event_ComponentEvent', function (e) {
});

Clazz.newMeth(C$, 'componentShown$java_awt_event_ComponentEvent', function (e) {
this.repaint();
});

Clazz.newMeth(C$, 'componentResized$java_awt_event_ComponentEvent', function (e) {
this.reinit();
this.repaint();
});

Clazz.newMeth(C$, 'actionPerformed$java_awt_event_ActionEvent', function (e) {
if (e.getSource() === this.swapButton ) {
var x;
for (x = 0; x < 2; x++) {
var y = this.vecs[0][x];
this.vecs[0][x] = this.vecs[1][x];
this.vecs[1][x] = y;
}
this.repaint();
}});

Clazz.newMeth(C$, 'adjustmentValueChanged$java_awt_event_AdjustmentEvent', function (e) {
});

Clazz.newMeth(C$, 'mouseDragged$java_awt_event_MouseEvent', function (e) {
this.edit$java_awt_event_MouseEvent(e);
});

Clazz.newMeth(C$, 'mouseMoved$java_awt_event_MouseEvent', function (e) {
if ((e.getModifiers() & 16) != 0) return;
this.edit$java_awt_event_MouseEvent(e);
});

Clazz.newMeth(C$, 'mouseClicked$java_awt_event_MouseEvent', function (e) {
});

Clazz.newMeth(C$, 'mouseEntered$java_awt_event_MouseEvent', function (e) {
});

Clazz.newMeth(C$, 'mouseExited$java_awt_event_MouseEvent', function (e) {
});

Clazz.newMeth(C$, 'mousePressed$java_awt_event_MouseEvent', function (e) {
if ((e.getModifiers() & 16) == 0) return;
var x = e.getX();
var y = e.getY();
var vc = Clazz.array(Integer.TYPE, [2]);
var i;
this.selection = -1;
var best = 900;
for (i = 0; i != 2; i++) {
this.findVecCoords$I$IA(i, vc);
var dx = x - vc[0];
var dy = y - vc[1];
var dist = dx * dx + dy * dy;
if (dist < best ) {
best = dist;
this.selection = i;
}}
if (this.selection != -1) this.edit$java_awt_event_MouseEvent(e);
});

Clazz.newMeth(C$, 'mouseReleased$java_awt_event_MouseEvent', function (e) {
if ((e.getModifiers() & 16) == 0) return;
this.selection = -1;
});

Clazz.newMeth(C$);
})();
//Created 2017-12-17 19:28:02
